### ISSUE TITLE
Global flags should be placed before the command name

### ISSUE THREAD

**Issue by kingaiakab** [mattallty/Caporal.js#21]
- `my_program --verbose test` should work
- `my_program test --verbose` should not work

In the first case `--verbose` is an option for `my_program`, in the second case it's an option for the `test` command.

**Comment by kinok** [comment]
Hello,

I disagree with this issue.
Many cli script helper (such as Symfony Console in php or Rails Console in RoR) work both way, and I think this is really convenient this way.

**Comment by deiucanta** [comment]
What about `git`?

- `git --version clone` shows the version & doesn't do the `clone` command
- `git clone --version` returns "error: unknown option `version`"

There are other well designed CLI apps the work this way. For me it's pretty obvious that the options are related to the first argument that precedes it.

Another example would be `docker` and `docker-compose`. You can have `docker-compose --project-name my_project up --timeout 5` - where the `--project-name` is a global option and the `--timeout` option is related to the `up` sub-command. 

A complex CLI app might also have overlapping option names that refer to different things (especially when it comes to the short notations of those options).

Just my 2 cents.

**Comment by mattallty** [comment]
It could be confusing for the user IMHO.
Putting options at the end, or at least always at the same place prevent user input errors, doesn't it ?

I love docker, but take this example:

If I first visit the `docker build` page:
https://docs.docker.com/engine/reference/commandline/build/
My first impression is that options come always *after* the command (build)

Then you go to this other page: (Use the Docker command line)
https://docs.docker.com/engine/reference/commandline/cli/
And you see the contrary :-/

So basically the correct synopsis for `docker build`would be:
* docker [OPTIONS] build [OPTIONS] [ARGS]

I'm not saying that they do things the wrong way, just wondering if it's the right thing to do.
And I like to privilege user experience, this does not seem to go this way (having to remember if a option is global or not, then putting it at the beginning or at the end, ...)

**Comment by mattallty** [comment]
Another thing: Even if we would consider options global or associated with command, it would no matter as you won't be proposing 2 options with the same name (it would be way too complicated for the end user). 

So allowing it before or after the command is just more convenient for the user.

**Comment by mattallty** [comment]
Closing for now
